### PR TITLE
Default datum transform for Server

### PR DESCRIPTION
**Pull request by mhugent**
(First travis test)

This PR adds the possibility to set default datum transformations for QGIS server (from environment variable or from QSettings). To achieve this, the following changes have been made:
- Reading of default datum transformation has been moved from QgsMapCanvas (gui) to QgsMapSettings (core) such that it can also be used by the server. If a dialog is needed (for the desktop case), a signal is sent from QgsMapSettings to QgsMapCanvas
- The server code has been changed from QgsMapRenderer to use the newer QgsMapRenderJob / QgsMapSettings classes

### CODE REVIEW

**Comment by m-kuhn** [comment]
Nice to see code ported from QgsMapRenderer to QgsMapRendererJob, that's certainly a win for the server.
I am not sure if making QgsMapSettings a QObject is a good idea. As far as I know QgsMapSettings should be a small overhead configuration object for a rendering job. I wonder what happens to the map settings object if it tries to find the datum transform in another thread than the main thread. I think QObjects have been designed with a private copy constructor, not sure what impacts overriding it has. And last I am also not sure about the overhead added by this. @wonder-sk what are your thoughts?

**Comment by mhugent** [comment]
The only reason for the QObject is the possibility to  send out a signal if there is no default datum transform (to show a dialog to the user in the GUI case). If there is another good possibility to achieve the same, them I'm of course fine with it.

Not sure if the overhead of QObject is really that big. And normally, we don't have thousands of QgsMapSettings 

>I wonder what happens to the map settings object if it tries to find the datum transform in another >thread than the main thread

Afaik signals/slots can be sent between threads without problems, see e.g. http://doc.qt.io/qt-4.8/threads-qobject.html: "More interesting is that QObjects can be used in multiple threads, emit signals that invoke slots in other threads, and post events to objects that "live" in other threads"

File: src/core/qgsmapsettings.cpp [hunk 1]
```diff
@@ -641,3 +676,79 @@ void QgsMapSettings::writeXML( QDomNode& theNode, QDomDocument& theDoc )
...
+      if ( slashSplit.at( 0 ) == srcAuthId && slashSplit.at( 1 ) == destAuthId )
+      {
+        datumTransformStore().addEntry( ml->id(), srcAuthId, destAuthId, slashSplit.at( 2 ).toInt(), slashSplit.at( 3 ).toInt() );
+        return;
+      }
+    }
+  }
+
+  //check if default datum transformation is available in settings
+  QSettings s;
```
Comment (NathanW2) [inline src/core/qgsmapsettings.cpp]: IMO `QgsMapSettings` shouldn't reach out to settings for things.. This makes it hard to use in standalone applications.   I'm not sure you would need a signal here.  Could it just be done with a different method `tranformRequired` that gets checked before doing this?

**Comment by m-kuhn** [comment]
On 04/12/2016 08:11 AM, mhugent wrote:
>
> The only reason for the QObject is the possibility to send out a
> signal if there is no default datum transform (to show a dialog to the
> user in the GUI case). If there is another good possibility to achieve
> the same, them I'm of course fine with it.
>

Couldn't this be checked when filling the QgsMapSetting with the
configuration?

> Not sure if the overhead of QObject is really that big. And normally,
> we don't have thousands of QgsMapSettings
>
>     I wonder what happens to the map settings object if it tries to
>     find the datum transform in another >thread than the main thread
>
> Afaik signals/slots can be sent between threads without problems, see
> e.g. http://doc.qt.io/qt-4.8/threads-qobject.html: "More interesting
> is that QObjects can be used in multiple threads, emit signals that
> invoke slots in other threads, and post events to objects that "live"
> in other threads"
>

Which connection type will be used? If it's Qt::AutoConnection it should
be mapped to a Qt::QueuedConnection which is posted on a different event
loop. In this case what happens to the reference which is sent with the
signal and supposed to be modified by the receiver?

> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/2989#issuecomment-208724040>
>

-- 
Matthias Kuhn
OPENGIS.ch - https://www.opengis.ch
Spatial • (Q)GIS • PostGIS • Open Source

**Comment by mhugent** [comment]
An alternative approach could be to still use QgsMapRenderer in the server and apply default datum transforms there. If it works, there will be no need for code changes in QgsMapSettings and other core classes (and also less changes in the server code).

**Comment by m-kuhn** [comment]
It would be nice to do this move, considering the comment

    // ### QGIS 3: remove QgsMapRenderer in favor of QgsMapRendererJob

https://github.com/qgis/QGIS/blob/master/src/core/qgsmaprenderer.h#L125

**Comment by wonder-sk** [comment]
Coming late here, but just for the record: +1 to @m-kuhn (QgsMapSettings should stay a simple data object, no signals/slots), and +1 to @NathanW2 (no QSettings in core classes)

**Pull request closed without merging**